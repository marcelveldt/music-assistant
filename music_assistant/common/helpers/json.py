"""Helpers to work with (de)serializing of json."""

from typing import Any

import orjson


JSON_ENCODE_EXCEPTIONS = (TypeError, ValueError)
JSON_DECODE_EXCEPTIONS = (orjson.JSONDecodeError,)


def json_encoder_default(obj: Any) -> Any:
    """Convert Special objects.

    Hand other objects to the original method.
    """
    if getattr(obj, "do_not_serialize", None):
        return None
    if (
        isinstance(obj, (list, set, filter, tuple))
        or obj.__class__ == "dict_valueiterator"
    ):
        return list(obj)
    if hasattr(obj, "as_dict"):
        return obj.as_dict()
    if isinstance(obj, bytes):
        return str(obj)

    raise TypeError


def json_dumps(data: Any) -> str:
    """Dump json string."""
    return orjson.dumps(
        data,
        option=orjson.OPT_NON_STR_KEYS | orjson.OPT_INDENT_2,
        default=json_encoder_default,
    ).decode("utf-8")


json_loads = orjson.loads
