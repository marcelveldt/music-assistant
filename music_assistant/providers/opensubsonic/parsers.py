"""Parse objects from py-opensonic into Music Assistant types."""

from __future__ import annotations

from typing import TYPE_CHECKING

from music_assistant_models.enums import ImageType
from music_assistant_models.media_items import Artist, MediaItemImage, ProviderMapping
from music_assistant_models.unique_list import UniqueList

if TYPE_CHECKING:
    from libopensonic.media import Artist as SonicArtist
    from libopensonic.media import ArtistInfo as SonicArtistInfo


def parse_artist(
    instance_id: str, sonic_artist: SonicArtist, sonic_info: SonicArtistInfo = None
) -> Artist:
    """Parse artist and artistInfo into a Music Assistant Artist."""
    artist = Artist(
        item_id=sonic_artist.id,
        name=sonic_artist.name,
        provider="opensubsonic",
        favorite=bool(sonic_artist.starred),
        provider_mappings={
            ProviderMapping(
                item_id=sonic_artist.id,
                provider_domain="opensubsonic",
                provider_instance=instance_id,
            )
        },
    )

    artist.metadata.images = UniqueList()
    if sonic_artist.cover_id:
        artist.metadata.images.append(
            MediaItemImage(
                type=ImageType.THUMB,
                path=sonic_artist.cover_id,
                provider=instance_id,
                remotely_accessible=False,
            )
        )

    if sonic_info:
        if sonic_info.biography:
            artist.metadata.description = sonic_info.biography
        if sonic_info.small_url:
            artist.metadata.images.append(
                MediaItemImage(
                    type=ImageType.THUMB,
                    path=sonic_info.small_url,
                    provider=instance_id,
                    remotely_accessible=True,
                )
            )

    return artist
