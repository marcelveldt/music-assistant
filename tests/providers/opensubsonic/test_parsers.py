"""Test we can parse Jellyfin models into Music Assistant models."""

import json
import pathlib

import aiofiles
import pytest
from libopensonic.media.artist import Artist, ArtistInfo
from syrupy.assertion import SnapshotAssertion

from music_assistant.providers.opensubsonic.parsers import parse_artist

FIXTURES_DIR = pathlib.Path(__file__).parent / "fixtures"
ARTIST_FIXTURES = list(FIXTURES_DIR.glob("artists/*.artist.json"))


@pytest.mark.parametrize("example", ARTIST_FIXTURES, ids=lambda val: str(val.stem))
async def test_parse_artists(example: pathlib.Path, snapshot: SnapshotAssertion) -> None:
    """Test we can parse artists."""
    async with aiofiles.open(example) as fp:
        artist = Artist(json.loads(await fp.read()))

    parsed = parse_artist("xx-instance-id-xx", artist).to_dict()
    # sort external Ids to ensure they are always in the same order for snapshot testing
    parsed["external_ids"].sort()
    assert snapshot == parsed

    # Find the corresponding info file
    example_info = example.with_suffix("").with_suffix(".info.json")
    async with aiofiles.open(example_info) as fp:
        artist_info = ArtistInfo(json.loads(await fp.read()))

    parsed = parse_artist("xx-instance-id-xx", artist, artist_info).to_dict()
    # sort external Ids to ensure they are always in the same order for snapshot testing
    parsed["external_ids"].sort()
    assert snapshot == parsed
