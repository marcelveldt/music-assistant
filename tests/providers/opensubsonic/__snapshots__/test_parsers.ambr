# serializer version: 1
# name: test_parse_artists[spec-artistid3.artist]
  dict({
    'external_ids': list([
    ]),
    'favorite': True,
    'item_id': '37ec820ca7193e17040c98f7da7c4b51',
    'media_type': 'artist',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': None,
      'explicit': None,
      'genres': None,
      'images': list([
        dict({
          'path': 'ar-37ec820ca7193e17040c98f7da7c4b51_0',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'thumb',
        }),
      ]),
      'label': None,
      'languages': None,
      'last_refresh': None,
      'links': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'release_date': None,
      'review': None,
      'style': None,
    }),
    'name': '2 Mello',
    'position': None,
    'provider': 'opensubsonic',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 0,
          'channels': 2,
          'content_type': '?',
          'output_format_str': '?',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': '37ec820ca7193e17040c98f7da7c4b51',
        'provider_domain': 'opensubsonic',
        'provider_instance': 'xx-instance-id-xx',
        'url': None,
      }),
    ]),
    'sort_name': '2 mello',
    'uri': 'opensubsonic://artist/37ec820ca7193e17040c98f7da7c4b51',
    'version': '',
  })
# ---
# name: test_parse_artists[spec-artistid3.artist].1
  dict({
    'external_ids': list([
    ]),
    'favorite': True,
    'item_id': '37ec820ca7193e17040c98f7da7c4b51',
    'media_type': 'artist',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': 'Empty biography',
      'explicit': None,
      'genres': None,
      'images': list([
        dict({
          'path': 'ar-37ec820ca7193e17040c98f7da7c4b51_0',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'thumb',
        }),
        dict({
          'path': 'http://localhost:8989/play/art/f20070e8e11611cc53542a38801d60fa/artist/2/thumb34.jpg',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': True,
          'type': 'thumb',
        }),
      ]),
      'label': None,
      'languages': None,
      'last_refresh': None,
      'links': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'release_date': None,
      'review': None,
      'style': None,
    }),
    'name': '2 Mello',
    'position': None,
    'provider': 'opensubsonic',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 0,
          'channels': 2,
          'content_type': '?',
          'output_format_str': '?',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': '37ec820ca7193e17040c98f7da7c4b51',
        'provider_domain': 'opensubsonic',
        'provider_instance': 'xx-instance-id-xx',
        'url': None,
      }),
    ]),
    'sort_name': '2 mello',
    'uri': 'opensubsonic://artist/37ec820ca7193e17040c98f7da7c4b51',
    'version': '',
  })
# ---
# name: test_parse_artists[spec-sample.artist]
  dict({
    'external_ids': list([
    ]),
    'favorite': True,
    'item_id': '100000002',
    'media_type': 'artist',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': None,
      'explicit': None,
      'genres': None,
      'images': list([
        dict({
          'path': 'ar-100000002',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'thumb',
        }),
      ]),
      'label': None,
      'languages': None,
      'last_refresh': None,
      'links': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'release_date': None,
      'review': None,
      'style': None,
    }),
    'name': 'Synthetic',
    'position': None,
    'provider': 'opensubsonic',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 0,
          'channels': 2,
          'content_type': '?',
          'output_format_str': '?',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': '100000002',
        'provider_domain': 'opensubsonic',
        'provider_instance': 'xx-instance-id-xx',
        'url': None,
      }),
    ]),
    'sort_name': 'synthetic',
    'uri': 'opensubsonic://artist/100000002',
    'version': '',
  })
# ---
# name: test_parse_artists[spec-sample.artist].1
  dict({
    'external_ids': list([
    ]),
    'favorite': True,
    'item_id': '100000002',
    'media_type': 'artist',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': 'Empty biography',
      'explicit': None,
      'genres': None,
      'images': list([
        dict({
          'path': 'ar-100000002',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'thumb',
        }),
        dict({
          'path': 'http://localhost:8989/play/art/f20070e8e11611cc53542a38801d60fa/artist/2/thumb34.jpg',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': True,
          'type': 'thumb',
        }),
      ]),
      'label': None,
      'languages': None,
      'last_refresh': None,
      'links': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'release_date': None,
      'review': None,
      'style': None,
    }),
    'name': 'Synthetic',
    'position': None,
    'provider': 'opensubsonic',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 0,
          'channels': 2,
          'content_type': '?',
          'output_format_str': '?',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': '100000002',
        'provider_domain': 'opensubsonic',
        'provider_instance': 'xx-instance-id-xx',
        'url': None,
      }),
    ]),
    'sort_name': 'synthetic',
    'uri': 'opensubsonic://artist/100000002',
    'version': '',
  })
# ---
